File: src/Memories/Domain/MemoryType.php
<?php

namespace App\Memories\Domain;

use App\Memories\Domain\Exceptions\InvalidType;
use App\Shared\Domain\Exceptions\CustomException;

class MemoryType
{
    public const LINK = 1;
    public const TEXT = 2;

    /**
     * @throws CustomException
     */
    private function __construct(public readonly int $value)
    {
        if ($value !== self::TEXT && $value !== self::LINK) {
            throw new InvalidType();
        }
    }

    public static function link(): self
    {
        return new self(self::LINK);
    }

    public static function text(): self
    {
        return new self(self::TEXT);
    }

    /**
     * @throws CustomException
     */
    public static function fromValue(int $value): self
    {
        return new self($value);
    }
}

File: src/main/kotlin/com/zaleort/second_brain/Memories/Domain/Memory/MemoryType.kt
package com.zaleort.second_brain.Memories.Domain.Memory

import com.zaleort.second_brain.Memories.Domain.Memory.Exceptions.InvalidMemoryTypeException

data class MemoryType(val value: Int) {
    init {
        if (value != LINK && value != TEXT) {
            throw InvalidMemoryTypeException("Invalid memory type: $value")
        }
    }

    override fun toString(): String {
        return when (value) {
            LINK -> "link"
            TEXT -> "text"
            else -> throw InvalidMemoryTypeException("Invalid memory type: $value")
        }
    }

    companion object {
        const val LINK = 1
        const val TEXT = 2

        fun link(): MemoryType {
            return MemoryType(LINK)
        }

        fun text(): MemoryType {
            return MemoryType(TEXT)
        }
    }
}